src/codecs/avif/decoder.rs

```diff
@@ -0,0 +1,156 @@
+            Ok(buf.len())
+        } else {
+            self.0.read_to_end(buf)
+        }
+    }
+}
+
+impl<'a, R: 'a + Read> ImageDecoder<'a> for AvifDecoder<R> {
+    type Reader = AvifReader<R>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        (self.picture.width(), self.picture.height())
+    }
+
+    fn color_type(&self) -> ColorType {
+        ColorType::Bgra8
+    }
+
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        let data = self.picture.plane_data_ptr(PlanarImageComponent::Y) as *mut _;
+        let (width, height) = self.picture.plane_data_geometry(PlanarImageComponent::Y);
+        let vec = unsafe { Vec::from_raw_parts(data, (width * height) as usize, (width * height) as usize) };
+        println!("into_reader");
+        Ok(AvifReader(Cursor::new(vec), PhantomData))
+    }
+
+    fn read_image(self, buf: &mut [u8]) -> ImageResult<()> {
+        assert_eq!(u64::try_from(buf.len()), Ok(self.total_bytes()));
+
+        use dcv_color_primitives as dcp;
+        dcp::initialize();
+
+        let buf = if self.picture.pixel_layout() != PixelLayout::I400 {
+            let pixel_format = match self.picture.pixel_layout() {
+                PixelLayout::I400 => todo!(),
+                PixelLayout::I420 => dcp::PixelFormat::I420,
+                PixelLayout::I422 => dcp::PixelFormat::I422,
+                PixelLayout::I444 => dcp::PixelFormat::I444,
+                PixelLayout::Unknown => panic!("Unknown pixel layout"),
+            };
+            let src_format = dcp::ImageFormat {
+                pixel_format,
+                color_space: dcp::ColorSpace::Bt601,
+                num_planes: 3,
+            };
+            let dst_format = dcp::ImageFormat {
+                pixel_format: dcp::PixelFormat::Bgra,
+                color_space: dcp::ColorSpace::Lrgb,
+                num_planes: 1,
+            };
+            let (width, height) = self.dimensions();
+            let src_buffers = &[
+                {
+                    let data = self.picture.plane_data_ptr(PlanarImageComponent::Y) as *mut _;
+                    let (width, height) = self.picture.plane_data_geometry(PlanarImageComponent::Y);
+                    unsafe { std::slice::from_raw_parts(data, (width * height) as usize) }
+                },
+                {
+                    let data = self.picture.plane_data_ptr(PlanarImageComponent::U) as *mut _;
+                    let (width, height) = self.picture.plane_data_geometry(PlanarImageComponent::U);
+                    unsafe { std::slice::from_raw_parts(data, (width * height) as usize) }
+                },
+                {
+                    let data = self.picture.plane_data_ptr(PlanarImageComponent::V) as *mut _;
+                    let (width, height) = self.picture.plane_data_geometry(PlanarImageComponent::V);
+                    unsafe { std::slice::from_raw_parts(data, (width * height) as usize) }
+                },
```

Comment:
The combination of `unsafe` and potentially overflowing multiplication `width * height` does _not_ bode well. Additionally, the unchecked cast with `as`? The length also seems to be reused several times over so validating it once initial by calculating the result with checked arithmetic would be preferred.

Reply:
Same here, latest version removed all `unsafe` blocks in favour of the `.plane()` API.